### PR TITLE
Add unit test for event persister sharding

### PR DESCRIPTION
**Pull request by erikjohnston**
This involves adding support in unit tests for:

1. using Redis for replication, this is done by creating a fake in memory one; and
2. allowing workers to send replication HTTP requests to other workers, rather than just master.

### CODE REVIEW

File: synapse/replication/tcp/redis.py [hunk 1]
```diff
@@ -228,3 +228,44 @@ def buildProtocol(self, addr):
...
+        convertNumbers,
+    )
+    factory.continueTrying = reconnect
+    for x in range(poolsize):
+        reactor.connectTCP(host, port, factory, connectTimeout)
+
+    if isLazy:
+        return factory.handler
+    else:
+        return factory.deferred
```
Comment (erikjohnston) [inline synapse/replication/tcp/redis.py:271]: This is a copy and paste from the function in `txredisapi` fwiw, but with `reactor` given as a parameter.

File: tests/unittest.py [hunk 2]
```diff
@@ -241,7 +241,7 @@ def setUp(self):
         # create a site to wrap the resource.
         self.site = SynapseSite(
             logger_name="synapse.access.http.fake",
-            site_tag="test",
+            site_tag=self.hs.config.server.server_name,
```
Comment (erikjohnston) [inline tests/unittest.py:244]: This is just a helpful thing to make logging in tests use the server name as the tag. Workers will use `<server_name>-<instance_name>`

File: tests/replication/test_sharded_event_persister.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        conf["stream_writers"] = {"events": ["worker1", "worker2"]}
+        conf["instance_map"] = {
+            "worker1": {"host": "testserv", "port": 1001},
+            "worker2": {"host": "testserv", "port": 1002},
+        }
+        return conf
+
+    def test_basic(self):
+        """Simple test that rooms can be created and joined when there are
+        multiple event persisters.
```
Comment (clokep) [inline tests/replication/test_sharded_event_persister.py:52]: There doesn't really seem to be any 'proof' in this that both event persisters are doing anything? I'm not convinced this is testing something.

File: tests/replication/_base.py [hunk 4]
```diff
@@ -197,19 +198,37 @@ def setUp(self):
...
 
-        self._worker_hs_to_resource = {}
+            self.hs.get_tcp_replication().start_replication(self.hs)
 
         # When we see a connection attempt to the master replication listener we
         # automatically set up the connection. This is so that tests don't
         # manually have to go and explicitly set it up each time (plus sometimes
         # it is impossible to write the handling explicitly in the tests).
+        #
+        # This sets registers the master replication listener:
```
Comment (clokep) [inline tests/replication/_base.py:227]:
```suggestion
        # Register the master replication listener:
```

Maybe? I'm not sure what this was trying to say.

File: tests/replication/_base.py [hunk 5]
```diff
@@ -467,3 +547,105 @@ def _run_once(self):
...
+
+    def send(self, msg):
+        """Send a message back to the client.
+        """
+        raw = self.encode(msg).encode("utf-8")
+
+        self.transport.write(raw)
+        self.transport.flush()
+
+    def encode(self, obj):
```
Comment (clokep) [inline tests/replication/_base.py:630]: This seems like something we should be able to pull from the txredis library or something?

File: tests/replication/_base.py [hunk 6]
```diff
@@ -467,3 +547,105 @@ def _run_once(self):
...
+            return "${len}\r\n{str}\r\n".format(len=len(obj), str=obj)
+        if isinstance(obj, int):
+            return ":{val}\r\n".format(val=obj)
+        if isinstance(obj, (list, tuple)):
+            items = "".join(self.encode(a) for a in obj)
+            return "*{len}\r\n{items}".format(len=len(obj), items=items)
+
+        raise Exception("Unrecognized type for encoding redis: %r: %r", type(obj), obj)
+
+    def connectionList(self, reason):
```
Comment (clokep) [inline tests/replication/_base.py:650]:
```suggestion
    def connectionLost(self, reason):
```

**Review by clokep — CHANGES_REQUESTED** [review]
Seems pretty reasonable, but I think the tests themselves could be a bit more robust?

File: tests/replication/test_sharded_event_persister.py (hunk 3 above)
Comment (erikjohnston) [reply tests/replication/test_sharded_event_persister.py:52]: Ish, sort of. It does It does cause the event to be persisted on a different worker and for it to come back via replication, and so it does it most code paths. TBH I was mostly using it as a way of testing the new code added to the base class.

File: tests/replication/_base.py (hunk 5 above)
Comment (erikjohnston) [reply tests/replication/_base.py:630]: You would have thought, but neither of the two redis libraries we use expose it :/

Comment (clokep) [reply tests/replication/_base.py:630]: That's unfortunate. 😢

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Looks good!

**Pull request merged**